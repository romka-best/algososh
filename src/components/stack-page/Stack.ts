import {IStack} from "./stack-page.types";

export default class Stack<T> implements IStack<T> {
    storage: T[] = [];

    constructor(private capacity: number = Infinity) {}

    push(item: T): void {
        if (this.size() === this.capacity) {
            throw Error("Stack has reached max capacity, you cannot add more items");
        }
        this.storage.push(item);
    }

    pop(): T | undefined {
        return this.storage.pop();
    }

    clear(): void {
        this.storage = [];
    }

    size(): number {
        return this.storage.length;
    }
}
