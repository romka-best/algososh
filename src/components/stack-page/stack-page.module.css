.root {

}

.form {
    display: flex;
    justify-content: center;
    gap: 12px;
}

.input {
    max-width: 378px;
}

.resetButton {
    margin-left: 68px;
}

.circles {
    display: flex;
    justify-content: center;
    gap: 16px;
    flex-wrap: wrap;

    margin-top: 22px;
}
