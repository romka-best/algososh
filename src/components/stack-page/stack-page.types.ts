import {ElementStates} from "../../types/element-states";

export interface IValue {
    value: string;
    type: ElementStates;
}
