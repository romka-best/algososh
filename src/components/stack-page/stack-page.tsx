import React from "react";

import {SolutionLayout} from "../ui/solution-layout/solution-layout";
import {Input} from "../ui/input/input";
import {Button} from "../ui/button/button";

import {SHORT_DELAY_IN_MS} from "../../constants/delays";

import {ElementStates} from "../../types/element-states";
import {IValue} from "./stack-page.types";

import styles from './stack-page.module.css';
import {Circle} from "../ui/circle/circle";

export const StackPage: React.FC = () => {
    const [isLoading, setIsLoading] = React.useState<boolean>(false);
    const [choiceAction, setChoiceAction] = React.useState<"Добавить" | "Удалить" | "Нет">("Нет");
    const [currentInputValue, setCurrentInputValue] = React.useState<string>("");
    const [stack, setStack] = React.useState<Array<IValue>>([]);

    const handleSubmitButton = React.useCallback((event: React.SyntheticEvent) => {
        event.preventDefault();

        if (isLoading) {
            return;
        }

        setIsLoading(true);
    }, []);
    const handleResetButton = React.useCallback((event: React.SyntheticEvent) => {
        event.preventDefault();

        setStack(clearStack());
    }, [clearStack]);

    React.useEffect(() => {
        if (isLoading) {
            const generatorAlgorithm = choiceAction === "Добавить" ?
                pushToStack(stack, {value: currentInputValue, type: ElementStates.Default}) :
                popFromStack(stack);
            setCurrentInputValue("");
            const interval = window.setInterval(() => {
                const generatorValue = generatorAlgorithm.next();

                setStack(() => {
                    const newState: Array<IValue> = [];
                    for (let i = 0; i < generatorValue.value.length; i++) {
                        newState.push(generatorValue.value[i])
                    }
                    return newState;
                });

                if (generatorValue.done) {
                    window.clearInterval(interval);
                    setIsLoading(false);
                }
            }, SHORT_DELAY_IN_MS);
        }
    }, [isLoading]);

    return (
        <SolutionLayout title="Стек">
            <div className={styles.root}>
                <form className={styles.form} onSubmit={handleSubmitButton} onReset={handleResetButton}>
                    <Input
                        extraClass={styles.input}
                        isLimitText={true}
                        maxLength={4}
                        disabled={isLoading}
                        value={currentInputValue}
                        onChange={(event: React.SyntheticEvent) => {
                            // @ts-ignore
                            setCurrentInputValue(String(event.target.value));
                        }}
                    />
                    <Button
                        text="Добавить"
                        disabled={isLoading || stack.length > 20 || !currentInputValue.length}
                        isLoader={isLoading && choiceAction === "Добавить"}
                        onClick={() => {
                            setChoiceAction("Добавить");
                        }}
                        type="submit"
                    />
                    <Button
                        text="Удалить"
                        disabled={isLoading || !stack.length}
                        isLoader={isLoading && choiceAction === "Удалить"}
                        onClick={() => {
                            setChoiceAction("Удалить");
                        }}
                        type="submit"
                    />
                    <Button
                        extraClass={styles.resetButton}
                        text="Очистить"
                        disabled={isLoading || !stack.length}
                        type="reset"
                    />
                </form>
                {stack.length || isLoading ? (
                    <div className={styles.circles}>
                        {
                            stack.map((value, index) => (
                                    <Circle
                                        key={index}
                                        head={index === stack.length - 1 ? "top" : undefined}
                                        tail={String(index)}
                                        letter={value.value}
                                        state={value.type}
                                    />
                                )
                            )
                        }
                    </div>
                ) : null}
            </div>
        </SolutionLayout>
    );
};

function* pushToStack(stack: Array<IValue>, item: IValue): Generator<Array<IValue>> {
    stack.push({
        ...item,
        type: ElementStates.Changing,
    });
    yield stack;

    stack[stack.length - 1] = {
        ...item,
        type: ElementStates.Default
    }
    return stack;
}

function* popFromStack(stack: Array<IValue>): Generator<Array<IValue>> {
    stack[stack.length - 1] = {
        ...stack[stack.length - 1],
        type: ElementStates.Changing,
    }
    yield stack;

    stack.pop();
    return stack;
}

function clearStack(): Array<IValue> {
    return [];
}
