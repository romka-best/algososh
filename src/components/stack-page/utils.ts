import Stack from './Stack';

import { ElementStates } from '../../types/element-states';

import { Step } from './stack-page.types';
import { OperationTypes } from './stack-page.types';

export function push<T>(value: T, list: Stack<T>) {
    const steps: Step<T>[] = [];

    steps.push({
        index: list.size(),
        value,
        list: list,
    });

    list.push(value);

    steps.push({
        list: list,
    });

    return steps;
}

export function pop<T>(list: Stack<T>) {
    const steps: Step<T>[] = [];

    const prevStack = new Stack<T>();
    prevStack.storage = Array.from(list.storage);

    steps.push({
        index: prevStack.size() - 1,
        list: prevStack,
    });

    list.pop();

    steps.push({
        list: list,
    });

    return steps;
}

export function clear<T>(list: Stack<T>) {
    const steps: Step<T>[] = [];

    steps.push({
        list,
    });

    list.clear();

    steps.push({
        list,
    });

    return steps;
}

export function getLetterState<T>(
    index: number,
    currentStep: Step<T>,
    operationType: OperationTypes | null,
): ElementStates {
    if (!operationType || currentStep.index === undefined) {
        return ElementStates.Default;
    }

    if (index === currentStep.index) {
        return ElementStates.Changing;
    }

    return ElementStates.Default;
}
