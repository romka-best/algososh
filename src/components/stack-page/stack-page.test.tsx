import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { StackPage } from './stack-page';

describe('StackPage', () => {
    describe('Отображение', () => {
        it('Начальное состяние', () => {
            const page = renderer
                .create(
                    <BrowserRouter>
                        <StackPage/>
                    </BrowserRouter>
                )
                .toJSON();

            expect(page).toMatchSnapshot();
        });
    });
});
