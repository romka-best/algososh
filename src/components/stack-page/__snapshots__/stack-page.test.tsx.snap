// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackPage Отображение Начальное состяние 1`] = `
<main
  className="content"
>
  <div
    className="titleBox"
  >
    <h1
      className="text text_type_h2 text_color_h1 title"
    >
      МБОУ АЛГОСОШ
    </h1>
    <span
      className="text text_type_fibonacci text_color_secondary subtitle"
    >
      им. Фибоначчи
    </span>
  </div>
  <div
    className="contentCard"
  >
    <a
      className="link"
      href="/"
      onClick={[Function]}
    >
      <button
        className="button returnButton"
        type="button"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M12.7073 4.29289C13.0978 4.68342 13.0978 5.31658 12.7073 5.70711L7.41436 11H19.0002C19.5524 11 20.0002 11.4477 20.0002 12C20.0002 12.5523 19.5524 13 19.0002 13H7.41436L12.7073 18.2929C13.0978 18.6834 13.0978 19.3166 12.7073 19.7071C12.3167 20.0976 11.6836 20.0976 11.293 19.7071L3.58594 12L11.293 4.29289C11.6836 3.90237 12.3167 3.90237 12.7073 4.29289Z"
            fill="#0032FF"
            fillRule="evenodd"
          />
        </svg>
        <p
          className="text text_type_button text_color_link ml-4"
        >
          К оглавлению
        </p>
      </button>
    </a>
    <h3
      className="text text_type_h3 text_color_h3 cardTitle"
    >
      Стек
    </h3>
    <div
      className="root"
    >
      <form
        className="form"
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <div
          className="content input"
        >
          <input
            className="input text text_type_input text_color_input"
            disabled={false}
            maxLength={4}
            onChange={[Function]}
            placeholder="Введите значение"
            type="text"
            value=""
          />
          <span
            className="text text_type_input-lim text_color_input mt-2 ml-8 limit"
          >
            Максимум — 4 символа
          </span>
        </div>
        <button
          className="text text_type_button text_color_primary button"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Добавить
          </p>
        </button>
        <button
          className="text text_type_button text_color_primary button"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          <p
            className="text"
          >
            Удалить
          </p>
        </button>
        <button
          className="text text_type_button text_color_primary button resetButton"
          disabled={true}
          onClick={[Function]}
          type="reset"
        >
          <p
            className="text"
          >
            Очистить
          </p>
        </button>
      </form>
    </div>
  </div>
  <p
    className="text text_type_column text_color_input mt-14 copyright"
  >
    © Сделано в Практикуме Романом Даниловым.
  </p>
</main>
`;
